[REVIEW] Adding some trivial `gdf_column`-related utility functions

eyalroz opened the pull request:
See issue #1387. Split off from PR #1366.

Note: The code was added, but nothing uses it that's added in this PR. But some code in #1366 does depend on it (e.g. code for resolving issue #1367). I'm PRing this way by request from @harrism .

jrhemstad commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,76 @@
+
+namespace cudf {
+
+gdf_error validate(const gdf_column& column)
```
Since these are new APIs and as of yet unused, I would prefer that they not use `gdf_error` codes and instead use `CUDF_EXPECTS` instead.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,105 @@
+
+} // namespace detail
+
+constexpr std::size_t inline size_of(gdf_dtype element_type) {
```
Redundant with https://github.com/rapidsai/cudf/blob/branch-0.7/cpp/src/dataframe/column.cpp#L200

I prefer your function, but ask that you consolidate with usage of `gdf_dtype_size`.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,105 @@
+    return type_dispatcher(element_type, detail::size_of_helper{});
+}
+
+inline std::size_t width(const gdf_column& col) { return size_of(col.dtype); }
```
Again, redundant with https://github.com/rapidsai/cudf/blob/branch-0.7/cpp/src/dataframe/column.cpp#L205 but I prefer your function. Please consolidate.

eyalroz replied:
Ooh, a chance to delete some code... Will do.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,105 @@
+// @TODO implement this using the type dispatcher
+// std::size_t data_size_in_bytes(const gdf_column& col);
+
+gdf_error validate(const gdf_column& column);
```
Most functions in this header I think are fairly self-explanatory, but these functions would benefit from Doxygen that describes what validating a column means.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,105 @@
+
+} // namespace detail
+
+bool are_strictly_type_compatible(
```
Doxygen describing this vs. the above `have_matching_types` would be useful.

eyalroz replied:
I unified the two.

kovaltan commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,83 @@
+    validate(*column_ptr);
+}
+
+bool have_matching_types(const gdf_column& validated_column_1, const gdf_column& validated_column_2) noexcept
```
Why doesn't this throw exceptions?
I often check input and output columns, usually dtype, size, null check is required, but somtimes dtype can be different.
How about adding this?

```
typedef enum
{
  DTYPE_CHECK = 0x000000001,
  SIZE_CHECK  = 0x000000002,
  NULL_CHECK  = 0x000000004,
  FULL_CHECK  = DTYPE_CHECK | SIZE_CHECK | NULL_CHECK,
} gdf_validate_flag;


void validate(const gdf_column& validated_column_1, const gdf_column& validated_column_2, gdf_validate_flag checkflag=FULL_CHECK)
{
  if( checkflag | DTYPE_CHECK )CUDF_EXPECTS ( (validated_column_1.dtype == validated_column_2.dtype) , "2 columns have different dtypes");
  if( checkflag | SIZE_CHECK )CUDF_EXPECTS ( (validated_column_1.size == validated_column_2.size) , "2 columns have different sizes");
  if( checkflag | NULL_CHECK )CUDF_EXPECTS ( (is_nullable(validated_column_1) == is_nullable(validated_column_2)) , "2 columns have different nulls");
}
```

eyalroz replied:
Do you mean `have_matching_types()`? It doesn't throw because it returns a boolean. It's a predicate.

As for vaidate() - I validate everything. Is it important to only make some of the checks?

jrhemstad commented:
@eyalroz there seems to be some unrelated files in this PR now. Can you make sure it's up to date and correct? Should be able to merge once reviews are addressed and the PR is corrected.

eyalroz commented:
@jrhemstad 
> @eyalroz there seems to be some unrelated files in this PR now. Can you make sure it's up to date and correct? Should be able to merge once reviews are addressed and the PR is corrected.
These were from another PR which this PR depended on; but the other PR was eventually committed, so they're not necessary. I rebased on branch-0.7, so now those extras are gone.

Note, however, that I _did_ make the type dispatch `constexpr auto` - and I needed to.

It should be ready to merge now.

harrism commented:
I'd still like to review this before merging.

jrhemstad commented on cpp/src/utilities/type_dispatcher.hpp:
```diff
@@ -115,7 +115,7 @@ namespace cudf {
 #pragma nv_exec_check_disable
 template <class functor_t, typename... Ts>
-CUDA_HOST_DEVICE_CALLABLE decltype(auto) type_dispatcher(gdf_dtype dtype,
+CUDA_HOST_DEVICE_CALLABLE constexpr auto type_dispatcher(gdf_dtype dtype,
```
The `decltype(auto)` is necessary to allow returning references https://isocpp.org/wiki/faq/cpp14-language#decltype-auto

eyalroz replied:
Done.

eyalroz commented:
@harrism :
> I'd still like to review this before merging.
Sure, go ahead.

jrhemstad commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,76 @@
+    const gdf_dtype_extra_info& rhs_extra_type_info) noexcept
+{
+    switch(common_dtype) {
+    case GDF_CATEGORY:  return lhs_extra_type_info.category == rhs_extra_type_info.category;
```
This check is problematic, and results from the hackiness in how we support the NVCategory column type.

`gdf_dtype_extra_info::category` is only set for columns of type `GDF_STRING_CATEGORY`. Furthermore, I don't think it is correct to check if two columns of type `GDF_STRING_CATEGORY` simply point to the same `NVCategory` object. Two category dictionaries can be compatible without pointing to the exact same NVCategory objects.

Presently, I do not believe there is any way to check if two `NVCategory` objects are compatible. Instead, in features like join, we explicitly synchronize the two NVCategory objects to guarantee they are compatible.

eyalroz replied:
What would you suggest, then? I have no preference one way or another here.

jrhemstad replied:
I would skip comparing `lhs_extra_type_info.category` for any column type.

eyalroz replied:
@jrhemstad : Ok, done.

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,76 @@
+bool have_same_type(const gdf_column& validated_column_1, const gdf_column& validated_column_2, bool ignore_extra_type_info) noexcept
+{
+    if (validated_column_1.dtype != validated_column_2.dtype) { return false; }
+    if (logical_xor(is_nullable(validated_column_1), is_nullable(validated_column_2))) { return false; }
```
Are columns really different types if one has a validity mask and the other does not? What if one is an input, and another is a transformation of that input that results in it having validity? These columns are now not the same type?

eyalroz replied:
Of course they aren't. In type-theoretic terms, one has type τ and the other has type τ ⊍ Unit (τ disjoint-union with a unit type). Not the same thing at all.

But types don't have to be the same for every single operator - as they are convertible; and we can have them implicitly conversion if we want. So the fact that the two columns in your example are not the same type should not be a problem.

harrism replied:
I'm arguing that this line (the logical_xor) should not be part of `have_same_type`.

harrism replied:
> So the fact that the two columns in your example are not the same type should not be a problem.

Then why is it part of the test?

eyalroz replied:
@harrism : Because, like I said, the two columns in your example are of different types. This is not a test of whether one type's domain is a strict superset of the other's / injective convertibility etc.

harrism replied:
I wrote:
> Are columns really different types if one has a validity mask and the other does not? 
To which you answered:
> Of course they aren't.
Then you wrote:
> the two columns in your example are of different types.

Hmmmm :) I don't need the type theory. I'm trying to determine if this function is actually useful as-is. I can imagine wanting to compare the dtype/dtypeinfo of two columns, ignoring whether they are nullable. Since you didn't put this function into use anywhere, it's hard to know whether it's useful as-is.

I don't like adding functions that aren't used.

eyalroz replied:
@harrism : Some API functions expect different argument columns to, well, have the same type. Right now it's a lot of functions, since we use out-parameters; but even if that decreases, there will still be a few.

I only wrote utility functions I was using myself in API function implementations. While its true that both my PRs for those (multisearch, apply boolean mask) have not yet been merged, these utilities are not specific to just those two.

eyalroz commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,119 @@
+/**
+ * @brief Returns the size in bytes of each element of a column (a.k.a. the column's width)
+ */
+inline std::size_t width(const gdf_column& col) noexcept { return size_of(col.dtype); }
```
The rationale here is that "size of" applies to gdf_dtypes, regardless of column.. It's the equivalent of C++ `sizeof()`. Also, for these scalar types - the meaning of "size" is obvious - the size in memory, in  bytes. For columns, however, the term "size", while idiomatic in C++ to denote the number of elements, is somewhat ambiguous - it could easily be interpreted as the size in bytes of the column memory representation. Now, a column is not a line: Like in an actual table on paper, it has two dimensions - a length and a width. I would actually support  using column _length_ rather than column _size_ everywhere, although that's a matter for a different PR / for the column redesign. The width is the fixed size of the element.

Of course, this metaphor doesn't hold perfectly for string columns, but for those, size_of is imperfect too.

harrism replied:
You seem to be agreeing that width and size_of do the same thing. Therefore one should be removed.

eyalroz replied:
But they don't do the same thing - one acts on columns and the other acts on types. Remember these are utility functions which are by definition non-essential. `is_nullable()` (now `has_validity()`) can always be replaced with a pointer nullptr comparison - but then it "sounds" different when you write the code. I want to be able to forget about the implementation details. Same thing for column widths. I want to be able to say "if these columns have the same width than blah blah", without explicitly referring to their type.

harrism replied:
Fine. Then back to my original concern: `width(column)` is vague. I think `size_of_dtype(column)` is clear. But that's because it is equally as clear as `size_of(column->dtype)`

I think if you are thinking about the width of the data in a column in bytes, you are OK with thinking about data types.

eyalroz replied:
@harrism : I'll drop the `width()` function for now. But I might bring this question up at some point later on, on nv-bar, to see whether people consider this vague or not.

eyalroz replied:
Following the discussion on Slack, will change this to `byte_width()` which AFAICT we can all live with. (Still think `width()` is more clear though.)

eyalroz commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,119 @@
+constexpr inline bool is_integral(const gdf_column& column) noexcept { return is_an_integer(column.dtype); }
+constexpr inline bool is_integral(const gdf_column* column) noexcept{ return is_integral(*column); }
+
+constexpr bool is_nullable(const gdf_column& column) noexcept { return column.valid != nullptr; }
```
To be honest, I think that "validity" is a much more ambiguous term than nullability. In the DBMS world, the terms "nullable" and "non-nullable" columns are very common and clearly understood - while "valid" can mean any number of things: Being accessible and usable, having been initialized, etc. In the context of libcudf, the potential for confusion has to do with pointers, and more specifically - pointers to gdf_column's. But - these will gradually go away, or at least the utility functions should not be used with pointers, but with gdf_columns proper (or their references). I've specifically refrained from accepting a gdf_column pointer for `is_nullable()`, and if you like I can drop the pointer-accepting versions of the utility functions from all over the file.

See also what I said about the different types. "has validity" is concerned with the implementation - the bitmask, which might be missing; `is_nullable()` is concerned with the column's element type.

eyalroz replied:
Oh, and I'll also mention the fact gdf_column's have a "null count", not an "invalid count"...

harrism replied:
cuDF is not a DBMS. Even the `data` column can have values that could be taken to mean "null" (e.g. NaN, zero, or empty string). "Nullable" is overloaded and confusing.

Agree that `null_count` should be renamed `invalid_count`.

eyalroz replied:
So, I disagree about which choice of name is better but I'll change this as you suggest.

eyalroz commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,76 @@
+ */
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "No column data despite there supposedly being rows.");
```
A column without data is fine. The problem is a column with no data that supposedly has elements. But - I shouldn't use the word "rows" here.

harrism replied:
I'm trying to make it less verbose. I think "Non-zero column size, but data pointer is null."

harrism replied:
I don't want error messages that use wishy-washy words like "supposedly" and "claims to". I want hard, explicit error messages.

Change it to: "Column has non-zero size but null data pointer."

eyalroz commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,76 @@
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type.");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column supposedly has null values, but no place to expose its record");
+    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column claims to have more null elements than elements overall");
```
"Invalid" is too vague. But will rephrase.

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+ */
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "Supposedly non-empty column is missing its data");
```
```suggestion
    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "Column size is non-zero but data pointer is null");
```

eyalroz replied:
The caller of this code does not need to know about the "data pointer". It's similar to "non-zero size"; I don't want to mention I have a size field. Minimum amount of necessary information...

harrism replied:
I don't want imprecise language like "supposedly" in error messages.

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "Supposedly non-empty column is missing its data");
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type");
```
```suggestion
    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Column dtype may not be GDF_invalid");
```

eyalroz replied:
I'll rephrase, buy I don't want to use the term "dtype" in a textual message

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "Supposedly non-empty column is missing its data");
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column claims to have null elements but has no validity mask to indicate element nullness");
```
```suggestion
    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column null_count is non-zero but valid pointer is null");
```

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column claims to have null elements but has no validity mask to indicate element nullness");
+    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column's null element count exceeds its overall number of elements");
```
```suggestion
    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column null_count greater than size");
```

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+#include <utilities/error_utils.hpp>
+
+// This should go up into more general utility code...
+static constexpr inline bool logical_xor(bool x, bool y) { return (x and (not y)) or ((not x) and y) ; }
```
This is completely unnecessary. `!=` serves the purpose of logical XOR in C++.

harrism commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+bool have_same_type(const gdf_column& validated_column_1, const gdf_column& validated_column_2, bool ignore_extra_type_info) noexcept
+{
+    if (validated_column_1.dtype != validated_column_2.dtype) { return false; }
+    if (logical_xor(is_nullable(validated_column_1), is_nullable(validated_column_2))) { return false; }
```
```suggestion
    if (is_nullable(validated_column_1) != is_nullable(validated_column_2)) { return false; }
```

eyalroz replied:
Ok, done, and - removed `logical_xor()` once its last use is gone.

harrism commented:
I had a discussion about the term "nullable" in our group, and found that it's fairly common in data science software (e.g. Arrow, Pandas) as well as databases. Therefore, I'm willing to accede on this point. If you would like to revert your change of `has_validity` to `is_nullable`, go for it.  I just have to get used to the term.

In the future we might want to disallow non-nullable columns, which would obsolete this function. 

Either way, in the redesign of cudf::column, we should probably rename the `valid` pointer to `nulls`, `is_null`, or something similar for consistency.

harrism commented:
Still many unaddressed comments (and the likely change *back* to is_nullable), so moving to 0.8 since 0.7 has one day left.

harrism commented:
@eyalroz do you want to proceed with this? Can you address remaining feedback?

eyalroz commented:
@harrism : Pushed changes which hopefully address all of your comments. Have not yet rebased on 0.8 - waiting for your approval.

eyalroz commented:
> @eyalroz do you want to proceed with this? Can you address remaining feedback?

Rebased on 0.8 and addressed - hopefully - the various suggestions, with the last commit. (I didn't "squash" all the commits, so there are a bunch of them; just the last one is really new here.)

harrism commented:
Please don't rebase a PR that is already under review. And please don't force push. Don't rebase anything that is public outside your personal fork, just use git merge. Please.

eyalroz commented:
So, I had interpreted the don't-rebase request to apply while on the same release branch and when there are outstanding review comments. I'll now expand it to "never after any review comments".

eyalroz commented:
So, I had interpreted the don't-rebase request to apply while on the same release branch and when there are outstanding review comments. I'll now expand it to "never after any review comments".

jrhemstad commented on cpp/src/column/column.cpp:
```diff
@@ -212,13 +211,17 @@ namespace{
 // returns the size in bytes of the specified gdf_dtype
-gdf_size_type gdf_dtype_size(gdf_dtype dtype) {
-  return cudf::type_dispatcher(dtype, get_type_size{});
+gdf_size_type gdf_dtype_size(gdf_dtype dtype)
```
Shouldn't this function be excised in favor of `cudf::size_of`?

eyalroz replied:
Yes it should. Although - there are places in which it's used like so: ``` gdf_dtype_size(my_gdf_column_.dtype)``` - and this I'm going to replace with `cudf::byte_width(my_gdf_column_)```.

harrism replied:
Why can't we excise API functions? We do that all the time? libcudf is not at 1.0 yet so we can change the API at will.

eyalroz replied:
@harrism : I suppose can, but why should it be in this PR, which is about adding utility functions? Perhaps someone might be using this function from Python, or have other objections?

harrism replied:
All it takes is a quick grep to see that this function is exposed in the Cython bindings but never called from the Python API: 
```
(cudf_dev) mharris@dgx14:~/RAPIDS/cudf$ grep -r gdf_dtype_size python/cudf/*                                                      python/cudf/bindings/cudf_cpp.pxd:    cdef gdf_size_type gdf_dtype_size(gdf_dtype dtype) except +
```

eyalroz replied:
@harrism : I had the (apparently mistaken) notion that if something was exposed through Cython then it was already part of the API and applications / cudf-using libraries might already be using it.

Would you guys like me to remove the two API functions?

harrism replied:
What two? I thought we were just talking about `gdf_dtype_size`. But yes, we should remove them -- that's the only way to prevent people using them. :)

eyalroz replied:
Done.

jrhemstad commented on cpp/src/column/column.cpp:
```diff
@@ -212,13 +211,17 @@ namespace{
 }
+
 // Returns the size in bytes of the data type of the gdf_column
 gdf_error get_column_byte_width(gdf_column * col, 
```
Likewise I'd think this should be excised in favor of `cudf::byte_width`

eyalroz replied:
Again, can't be done since it's an API function :-(

I know making the API C++'ish is decided, but actually - it seems we haven't considered changes such as exposing some of the `namespace cudf` functionality outwards.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,119 @@
+template<typename T>
+T* get_data(const gdf_column* column) noexcept { return get_data<T>(*column); }
+
+constexpr inline bool is_an_integer(gdf_dtype element_type)
```
I'd prefer that this kind of thing go through the `type_dispatcher` to limit the amount of code that need change when new types are added/modified, e.g., if/when we add unsigned integral types.

Something like this should do it:

```
struct integral_check{
   template <typename T>
   constexpr bool operator()(void) const noexcept{
      return std::is_integral<T>::value;
   }
};

...

return cudf::type_dispatcher(element_type, integral_check{});
```

eyalroz replied:
Why would you prefer using a complex and not-narrattively flowing gadget for this kind of a function? As for using `std::is_integral` - we could use that, but then I'd need the inverse function of cudf::size_of (i.e. `cudf::type_of`).

Actually, a `cudf::type_of` might be nice, because we could use it in the implementation type_dispatcher as well.

Also, using `std::integral` requires that we guarantee that only GDF_INTXX or GDF_UINTXX types will use an unwrapped integer type, and vice versa, that no type which "is an integer" will ever be wrapped.

If you're willing to make that guarantee, I'll switch to your suggestion or use `cudf::type_of` if you approve of it.

jrhemstad replied:
I think I wasn't clear.

I'm not suggesting getting rid of `is_an_integer`. Instead, my suggestion was to update the implementation of `is_an_integer`. 

For example:

```
struct integral_check{
   template <typename T>
   constexpr bool operator()(void) const noexcept{
      return std::is_integral<T>::value;
   }
};
constexpr inline bool is_an_integer(gdf_dtype element_type){
   return cudf::type_dispatcher(element_type, integral_check{});
}
```

eyalroz replied:
You were clear... I'm just questioning whether it's an upgrade. It's more convoluted and obscure (as much as I like our type dispatcher), and requires making the implicit assumption I mentioned.

jrhemstad replied:
> Why would you prefer using a complex and not-narrattively flowing gadget for this kind of a function?

Because then when we add `GDF_UINT*` I have to remember to go back and update this function. Ideally, I want to keep all of the changes required when a new type is added or changed to be limited to as few places as possible. 

> cudf::type_of

Making sure I understand, you're suggesting a trait that maps `gdf_dtype` to concrete C++ types? If so, I support that and updating `type_dispatcher` to use the trait.

That said, I don't know how that'd help the implementation of `is_an_integer`, you'd still need a switch statement over `element_type`. That's exactly the kind of thing I'm trying to minimize. 

> Also, using std::integral requires that we guarantee that only GDF_INTXX or GDF_UINTXX types will use an unwrapped integer type, and vice versa, that no type which "is an integer" will ever be wrapped.

I feel comfortable making that guarantee for now. If we _do_ need to use a wrapper instead, we can always overload `std::is_integral<T>` for user-defined types.

eyalroz replied:
Given your guarantee, let's not split hairs and I'll go with your suggestion.

jrhemstad commented on cpp/src/utilities/column_utils.cpp:
```diff
@@ -0,0 +1,75 @@
+
+bool have_same_type(const gdf_column* validated_column_ptr_1, const gdf_column* validated_column_ptr_2, bool ignore_extra_type_info) noexcept
+{
+    return have_same_type(*validated_column_ptr_1, *validated_column_ptr_2);
```
Check for the column points being `nullptr`.

eyalroz replied:
But the have_same_type functions assume their arguments have already been validated. Says so right in the doxygen comment... are you sure?

eyalroz commented:
@jrhemstad : Pushed the changes you requested (but have not removed any C API function; I think that's a matter for a different PR).

harrism commented:
Hmmm, `test_concat_all_nulls` pytest is failing?

eyalroz commented:
> Hmmm, `test_concat_all_nulls` pytest is failing?

Not for me:
```
[ 56%] cudf/tests/test_multi.py::test_concat[False-none] PASSED
[ 56%] cudf/tests/test_multi.py::test_concat[False-some] PASSED
[ 56%] cudf/tests/test_multi.py::test_concat[False-all] PASSED 
[ 56%] cudf/tests/test_multi.py::test_concat[z-none] PASSED    
[ 56%] cudf/tests/test_multi.py::test_concat[z-some] PASSED    
[ 56%] cudf/tests/test_multi.py::test_concat[z-all] PASSED     
[ 56%] cudf/tests/test_multi.py::test_concat[y-none] PASSED    
[ 56%] cudf/tests/test_multi.py::test_concat[y-some] SKIPPED   
[ 56%] cudf/tests/test_multi.py::test_concat[y-all] SKIPPED    
[ 56%] cudf/tests/test_multi.py::test_concat_all_nulls[values0] PASSED              
[ 56%] cudf/tests/test_multi.py::test_concat_all_nulls[values1] PASSED              
[ 56%] cudf/tests/test_multi.py::test_concat_all_nulls[values2] PASSED              
[ 56%] cudf/tests/test_multi.py::test_concat_errors PASSED     
[ 56%] cudf/tests/test_multi.py::test_concat_misordered_columns PASSED    
```

jrhemstad commented:
rerun tests

shwina commented:
This failure was fixed in https://github.com/rapidsai/cudf/pull/1780. I think that merging branch-0.8 into this PR's branch should take care of this

harrism commented:
@eyal please merge branch-0.8 into your branch and resolve conflicts so we can proceed.

eyalroz commented:
@harrism : Merged.

jrhemstad commented:
rerun tests

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,127 @@
+namespace cudf {
+
+template<typename T>
+T* get_data(const gdf_column& column) noexcept { return static_cast<T*>(column.data); }
```
Avoid ODR violation, and it'd be nice for this to work in device code. 
```suggestion
__host__ __device__
inline T* get_data(const gdf_column& column) noexcept { return static_cast<T*>(column.data); }
```

eyalroz replied:
I did indeed forget to inline.

However, these are strictly-host-side utilities; which is why it's an `.hpp` rather than a `.cuh` file. There are generally no `gdf_column`s on the device side for now AFAICR, so let's not go there.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,127 @@
+T* get_data(const gdf_column& column) noexcept { return static_cast<T*>(column.data); }
+
+template<typename T>
+T* get_data(const gdf_column* column) noexcept { return get_data<T>(*column); }
```
```suggestion
__host__ __device__
inline T* get_data(const gdf_column* column) noexcept { return get_data<T>(*column); }
```

eyalroz replied:
Same comment as above.

jrhemstad commented on cpp/src/utilities/column_utils.hpp:
```diff
@@ -0,0 +1,127 @@
+constexpr inline bool is_integral(const gdf_column& column) noexcept { return is_an_integer(column.dtype); }
+constexpr inline bool is_integral(const gdf_column* column) noexcept{ return is_integral(*column); }
+
+constexpr bool is_nullable(const gdf_column& column) noexcept { return column.valid != nullptr; }
```
Could you also add a utility function like `has_nulls` that first checks if the column `is_nullable` and then returns if `null_count > 0`?

eyalroz replied:
I don't know about that. I mean, if the column is not known to be valid, just `validate()` it when you get it. If it is, `null_count > 0` works. But - you can convince me this would be a useful addition.

jrhemstad replied:
> If it is, null_count > 0 works.

True, but `col->valid != nullptr` also just works in place of `is_nullable(col)`. 

The point is to abstract away the implementation details so that you don't have to think about them and provide functions that contribute to a nicer overall narrative flow. 

As far as naming, I don't think it's ambiguous.

```
is_nullable(col) // this means a column _can_ contain nulls
has_nulls(col) // this means a column _does_ contain nulls
```

eyalroz replied:
@jrhemstad : Well, if you think people will use it, then ok.

jrhemstad replied:
Just `const&` is fine with me.

jrhemstad commented on python/cudf/bindings/cudf_cpp.pxd:
```diff
@@ -222,9 +222,14 @@ cdef extern from "cudf.h" nogil:
-    cdef gdf_size_type gdf_dtype_size(gdf_dtype dtype) except +
-
-    cdef gdf_error get_column_byte_width(gdf_column * col, int * width) except +
+    cdef gdf_error gdf_filter(size_t nrows,
```
`gdf_filter` was removed. It looks like this was mistakenly added back during a merge. You can remove the `gdf_filter` from this file.

eyalroz commented:
@jrhemstad : So, are we done?

harrism commented:
@kkraus14 @thomcom or someone from python-codeowners: can you please review the trivial change to cython bindings here?
